cEP-0022.md

```diff
@@ -0,0 +1,162 @@
+
+- Files having similar kind of names can be having different codestyles.
+- Files in a given directory can be following different codestyles.
+- Different parts of files maybe even following a different codestyle depending
+  on function name which can also be handled by placing `ignore`s at
+  appropriate places in the file dict, such kind of functionality can only be
+  added to the bears themselves, but instead we stop our train of thought over
+  here.
+
+## Learning from real world projects
+
+We save ourselves the additional work for building some feature that may
+hardly ever be used by any org, instead we stop assumming and start
+implementing and learning from our available resources.
+
+We now choose a list of orgs for which we try to test our `green-mode`
+ourselves. We check whether brute force is taking too much time or whether
+the `lite-mode` is producing the correct results (i.e. *green* results). We
+choose these orgs in such a way that they are very famous or very well known or
+act as upstream repositories for a huge amount of orgs. This way a very large
+number of orgs may be mimicking the code style of these orgs itself and if
+our further optimizations can fix the problem of generating *green* config
+files for these organizations, we are in turn solving the problem for these
+other orgs/communities at the same time.
+
+We learn manually from these orgs, what combinations of these bear-setting
+values are they using, whether they are using it all over their project or only
+in certain scenarious, whether they have different codestyles in different
+directories or different nomenclature of file patterns require a different
+set of settings in their project. We give `coala-quickstart`, the ability
+to learn and identify these patterns, recognizing these scenarios for further
+runs.
+
+# Developing tools for a better future
+
+`coala-quickstart` should also develop custom logs which make further
+developing and identifying the cause of our config files to generate errors
+more evident and easier to grasp for further development. It is possible that
+we may have to apply some other set of rules if we identify a regex pattern.
+The regex pattern identification will have to be done manually but once the regex
+is formed, and provided we find this a problem in our chosen orgs, helper
+functions/classes should be developed for the bears so that they are able to
+put `ignore`s at correct places in the `file_dict` on identifying the regex and
+use different bear settings for different portions of the file itself.
+**(Stretch Goal)**
+
+# Engaging newcomers to spread the word about coala
+
+We can only look at a finite number of orgs within the coding period so the
+last few weeks of the project should deal with writing docs and adding newcomer
+tasks for either opening PRs in other orgs of the *green* config files or
+notifying us of the results if they are not able to generate one and opening
+issues for further improvement.
```

Comment:
Not sure if that would be a good idea, but we can script the process of generating a green mode coafile, creating a PR for that to the projects (using corobo's account) with a human-like but templated message.

Reply:
Automated, unsolicited PRs are a big no-no according to Github's terms. It should be by a human.

Besides, we need to follow that org's commit guidelines just like we expect newcomers to follow coala's.

Reply:
I propose we leave this out of the gsoc project and make it a GCI task. Repeatable, isolated, and moderately difficult; it's kinda perfect.

CC @jayvdb

Reply:
This items mentions

- "should deal with writing docs" which will be needed, and
- "adding newcomer tasks", which I agree should be instead aimed at GCI, so this should instead be an overhaul of https://github.com/coala/projects/blob/master/_projects/use_coala.md and `use_coala_2.md` and `use_coala_3.md` to use green mode, and test them on a few various types of projects, so that have better tasks for the next GCI.